### ISSUE TITLE
NetIncomingMessage.SenderConnection occasionally null

### ISSUE THREAD

**Issue by jbruening** [lidgren/lidgren-network-gen3#45]
SenderConnection occasionally evaluates to null, and then suddenly isn't

![senderconnection](https://cloud.githubusercontent.com/assets/1760487/10114647/726d3b96-63a9-11e5-8d26-627888e81349.PNG)

built off c42b885

**Comment by jbruening** [comment]
I guess for more information if it helps:

msg.Data was a valid byte array with 20215 bytes
delivery method was UnreliableSequenced
LengthBits was 161720

**Comment by lidgren** [comment]
The code you pasted isn't part of the library... do you get this problem with the samples as well?

**Comment by jbruening** [comment]
I have a simple program that can reproduce this issue

http://pastebin.com/0Zn8nqXV

Just make a new project with a reference to a build from c42b885

This should also reproduce #43, but it doesn't seem to be. Increasing the time between clients connecting to something like 10 seconds might reproduce it.

**Comment by jbruening** [comment]
Definitely a threading issue. Using the code in the pastebin, I put a few more variables to peek at the state of the incoming message before breaking.

![threading](https://cloud.githubusercontent.com/assets/1760487/10149916/75a58922-65f1-11e5-8c1d-c05cd6af3de1.PNG)

The type also switched in that image from a Data to an Error, which I'm assuming is the reason for the nulls.

**Comment by jbruening** [comment]
That commit doesn't seem to entirely fix the problem though.

**Comment by lidgren** [comment]
Can you verify setting UseMessageRecycling to false suppressed the problem?

**Comment by jbruening** [comment]
The program no longer works - 'RespondedConnect' messages are constantly received on the server, with the position of the message past the end. Setting the message's position to 0 does not fix it, as again, that message is constantly received.
![out of bounds](https://cloud.githubusercontent.com/assets/1760487/10171172/99dffe0c-668d-11e5-863c-1fb9675f6ae3.PNG)

**Comment by jbruening** [comment]
Shouldn't have deleted it, it's definitely still happening - 

The program stops working. The server starts getting statuschanged messages that are at the end of their array.
![out of bounds](https://cloud.githubusercontent.com/assets/1760487/10171693/3166f38c-6690-11e5-9376-bbe28bc45db3.PNG)
Attempting to change the position to 0 does not help, as the messages are still constantly received.

**Comment by lidgren** [comment]
Did you try UseMessageRecycling false? I also advice running in Debug as more checks are made there. I presume you are not manually recycling the message before it's read.

**Comment by jbruening** [comment]
yes, that last message is with UseMessageRecycling set to false.

Debug build throws on an assert about it being out of bounds, before it attempts to do the read, but it's still effectively doing the same thing.

The chat example runs fine for me

**Comment by jbruening** [comment]
This seems to be stemming from the UnreliableSizeBehaviour being set to NormalFragmentation.

**Comment by lidgren** [comment]
Aha. UnreliableSizeBehaviour.NormalFragmentation is a very suspect feature kept mostly for backwards compability. It should never really be used as it stands right now.
I'm hoping to implement fragmentation for unreliable messages better (some kind of timeout for collecting incomplete fragments) some day...
